network/p2p/p2pnode/libp2pNode_test.go

```diff
@@ -384,6 +384,63 @@ func TestCreateStream_SinglePeerDial(t *testing.T) {
 	require.Equal(t, expectedCreateStreamRetries, createStreamRetries.Load(), fmt.Sprintf("expected %d dial peer retries got %d", expectedCreateStreamRetries, createStreamRetries.Load()))
 }
 
+// TestCreateStream_InboundConnResourceLimit ensures that the setting the resource limit config for PeerDefaultLimits.ConnsInbound restricts the number of inbound
+// connections created from a peer to the configured value.
+func TestCreateStream_InboundConnResourceLimit(t *testing.T) {
+	idProvider := mockmodule.NewIdentityProvider(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	signalerCtx := irrecoverable.NewMockSignalerContext(t, ctx)
+
+	sporkID := unittest.IdentifierFixture()
+
+	uniMgrFactory := p2ptest.UnicastManagerFixtureFactory()
+
+	sender, id1 := p2ptest.NodeFixture(
+		t,
+		sporkID,
+		t.Name(),
+		p2ptest.WithConnectionGater(testutils.NewConnectionGater(idProvider, func(pid peer.ID) error {
+			return nil
+		})),
+		p2ptest.WithDefaultResourceManager(),
+		p2ptest.WithUnicastManagerFactoryFunc(uniMgrFactory),
+		p2ptest.WithCreateStreamRetryDelay(10*time.Millisecond))
+
+	receiver, id2 := p2ptest.NodeFixture(
+		t,
+		sporkID,
+		t.Name(),
+		p2ptest.WithDefaultResourceManager(),
+		p2ptest.WithConnectionGater(testutils.NewConnectionGater(idProvider, func(pid peer.ID) error {
+			return nil
+		})),
+		p2ptest.WithCreateStreamRetryDelay(10*time.Millisecond))
+
+	idProvider.On("ByPeerID", sender.Host().ID()).Return(&id1, true).Maybe()
+	idProvider.On("ByPeerID", receiver.Host().ID()).Return(&id2, true).Maybe()
+
+	p2ptest.StartNodes(t, signalerCtx, []p2p.LibP2PNode{sender, receiver}, 100*time.Millisecond)
+	defer p2ptest.StopNodes(t, []p2p.LibP2PNode{sender, receiver}, cancel, 100*time.Millisecond)
+
+	pInfo, err := utils.PeerAddressInfo(id2)
+	require.NoError(t, err)
+	sender.Host().Peerstore().AddAddrs(pInfo.ID, pInfo.Addrs, peerstore.AddressTTL)
+
+	var wg sync.WaitGroup
+	for i := 0; i < 20; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			_, err = sender.CreateStream(ctx, receiver.Host().ID())
+			require.NoError(t, err)
+		}()
+	}
```

Comment:
Please document that the purpose of this loop is not just to create streams, but also to drive the libp2p protocol towards a state where it attempts to establish concurrent connections, which should be safeguarded by the single inbound connection guarantee.

Reply:
https://github.com/onflow/flow-go/pull/4022/commits/3e89bdabced0cf4c353cedb814e1197164560bdc